### ISSUE TITLE
recreate stops making progress

### ISSUE THREAD

**Issue by lexelby** [duplicati/duplicati#4585]
- [x] I have searched open and closed issues for duplicates.
- [x] I have searched the [forum](https://forum.duplicati.com) for related topics.

----------------------------------------

## Environment info

- **Duplicati version**: 2.0.6.3-1
- **Operating system**: Ubuntu 16.04.5 LTS
- **Backend**: rclone (v1.54.0)

## Description

I was seeing some weird behavior, so I decided to try recreating the database.  I've run it 3 times now and had similar results: in every case, somewhere along the line, duplicati stops making progress.  I don't mean it's "taking forever" -- it's doing nothing.  No CPU activity, no I/O activity (per iostat), timestamp on the sqlite file is not changing, and the live log isn't showing any new messages.  Last message looks like this:

```
ExecuteScalarInt64: SELECT "ID" FROM "FileLookup" WHERE "PrefixID" = 32239 AND "Path" = "redacted" AND "BlocksetID" = 41956 AND "MetadataID" = 259049 took 0:00:00:00.000
```

That's about 50 minutes ago.

It stops at a different point each time.  Once it got through almost all of my dlist files.  Waiting hours seems to make no difference (the most I wanted was 10).

Each time, the only way I have found to move on is to restart the duplicati daemon and then start another recreate.

**Comment by lexelby** [comment]
I ran a recreate on the command line, and got the same result.  Here's the log up to when it stopped making progress:

```
Processing filelist volume 51 of 127
Backend event: Get - Started: duplicati-20210515T070001Z.dlist.zip (64.93 MB)
  Downloading file (64.93 MB) ...
Backend event: Get - Completed: duplicati-20210515T070001Z.dlist.zip (64.93 MB)
Processing filelist volume 52 of 127
Backend event: Get - Started: duplicati-20210516T070001Z.dlist.zip (65.49 MB)
  Downloading file (65.49 MB) ...
Backend event: Get - Completed: duplicati-20210516T070001Z.dlist.zip (65.49 MB)
Processing filelist volume 53 of 127
Backend event: Get - Started: duplicati-20210517T070001Z.dlist.zip (64.93 MB)
  Downloading file (64.93 MB) ...
```

I was continually running `ps` in the background and grepping for rclone processes launched by duplicati.  I saw this:

```
/usr/bin/rclone copyto gcs:lexelby-duplicati/backup/duplicati-20210517T070001Z.dlist.zip local:/tmp/dup-29d192c2-a27e-4723-9074-38502f780a5a
```

That process finished and the file `/tmp/dup-29d192c2-a27e-4723-9074-38502f780a5a` existed but was 0 bytes.  So maybe we're looking at some kind of bug around error handling for the rclone backend?

Why not use the gcs storage backend?  I'm using the rclone backend so that I can pass a bwlimit option and get finer-grained, time-based control of bandwidth usage.

I tried jump-starting duplicati by re-running the `rclone` command above.  It downloaded successfully, but unsurprisingly, duplicati didn't notice.

**Comment by warwickmm** [comment]
Thanks @lexelby.  We've had [numerous reports](https://github.com/duplicati/duplicati/issues/4041) of database recreation performance issues.  [Some](https://forum.duplicati.com/t/recreating-database-logic-understanding-issue-slow/6496/70) were [fixed](https://github.com/duplicati/duplicati/pull/3758), but there may be other areas of improvement.  There is a mention [here in the forums](https://forum.duplicati.com/t/recreating-database-hanging-for-days/12707/5) of a slow query (consuming very little CPU, which can appear like nothing is happening) that may be related.

The reasons for slow recreation can vary depending on the situation.  I recommend that you create a post in the forums so that we can better diagnose your issue.  There are many more users there that may be able to help.

I'm also going to close this in favor of issue #4041, which documents many other cases of poor recreate performance.

**Comment by warwickmm** [comment]
I just noticed your follow-up comment.  Is it consistently getting stuck on a file download?  Please include this detail, as well as some information on your configuration (block size, remote volume size, size of source files, number of backup versions, etc.) in your forum post.

**Comment by warwickmm** [comment]
https://github.com/duplicati/duplicati/issues/4328 could be a related issue that concerns rclone hanging.

**Comment by lexelby** [comment]
Thanks!  That definitely seems promising, and it explains why I couldn't find any other reports.  Seems like I have a couple of options to fix:

* try --disable-piped-streaming (and maybe turn that off for restores...?)
* maybe --number-of-retries might help?
* maybe --rclone-option=--retries=10 might help (running that test now)
* switch to GCS backend and figure out my throttling in some other way

> I just noticed your follow-up comment. Is it consistently getting stuck on a file download?

Now that I know to look, I'll keep an eye out on my current and subsequent tests.

> Please include this detail, as well as some information on your configuration (block size, remote volume size, size of source files, number of backup versions, etc.) in your forum post.

I'd really rather not create yet another forum account, especially since it sounds like #4328 is a likely culprit.  My source files are around 400GB, backend file size is not much more than that, 127 backup versions.

**Comment by lexelby** [comment]
Here's my current recreate attempt:

```
duplicati-cli repair rclone://gcs/lexelby-duplicati/backup?rclone-local-repository=local --backup-name=backup --backup-test-percentage=0 --backup-test-samples=0 --compression-module=zip --dblock-size=1gb --dbpath=redacted/.config/Duplicati/LJEFKNFJPP.sqlite --disable-module=console-password-input --no-auto-compact --no-backend-verification --no-encryption --log-level=Verbose --rclone-option=--retries=10
```

1GB file blocks are because I'm using GCS archival storage and it gets expensive to do a lot of API calls.

**Comment by warwickmm** [comment]
Typically, "archival" class storage is problematic.  However, it sounds like GCS's archival storage is different in that your files are supposed to be available within milliseconds (vs. possibly hours or days).

> switch to GCS backend and figure out my throttling in some other way

It would be very helpful if in your testing you can confirm if the issue is related to #4328.  Switching to GCS backend could reveal 
 this.

> 1GB file blocks are because I'm using GCS archival storage and it gets expensive to do a lot of API calls.

Is this the [block size](https://www.duplicati.com/articles/Choosing-Sizes/#the-block-size), or [remote volume size](https://www.duplicati.com/articles/Choosing-Sizes/#remote-volume-size)?  If it's the block size, that's probably way too large and can be probelmatic.  For 400GB of source files, that's quite large for the remote volume size as well.  I would again recommend getting help in the forums to get suggestions on the best configuration for your situation.   

https://forum.duplicati.com/

**Comment by lexelby** [comment]
Uh oh.  Now I'm not so sure.  I thought I was setting the remote volume size to 1GB but I might've screwed that up and set the block size instead!  I'll have to verify.  Maybe it's time to start over.

**Comment by lexelby** [comment]
Looks like I'm referring to the "dblock-size" option, which is this:

> This option can change the maximum size of dblock files. Changing the size can be useful if the backend has a limit on the size of each individual file
> Default value: "50mb"

So I think that's the volume size.

> It would be very helpful if in your testing you can confirm if the issue is related to #4328. Switching to GCS backend could reveal this.

Got it.  I'll do that for my next trial.

**Comment by lexelby** [comment]
Ah, I remember now.  I didn't choose rclone because of the bandwidth limiting options, I chose it because the GCS backend's oauth implementation scares me.  I don't think I'd feel comfortable using it unless it supported a service account JSON key like rclone.

**Comment by warwickmm** [comment]
Understood.  In case you haven't seen it, here's a description of how Duplicati works with OAuth:

https://www.duplicati.com/articles/OAuth/

**Comment by lexelby** [comment]
My attempt with `--rclone-option=--retries=10` failed in the same way: rclone produced a 0-length file and duplicati locked up.  I decided to try the s3-compatible backend (sure wish I'd noticed that one ages ago...). It's currently happily trucking along, downloading dlists from my GCS bucket.  If this works, I think this could provide similar confirmation to what you were hoping for if I switched to the GCS backend.
